### PR TITLE
[NEXUS-7654] Use NFC instead of attributes for attribute checksum "not-found"

### PR DESCRIPTION
**Pull request by jdillon**
https://issues.sonatype.org/browse/NEXUS-7654

Attempts to use the NFC to determine "not-found"-ness of a checksum request, instead of attributes.

### CODE REVIEW

**Comment by jdillon** [comment]
http://bamboo.s/browse/NX-OSSF628-1 :white_check_mark:

**Comment by adreghiciu** [comment]
Shouldn't we remove the hash from NFC if we found it?

File: components/nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/ChecksumContentValidator.java [hunk 1]
```diff
@@ -194,37 +184,32 @@ private static RemoteHashResponse doRetrieveChecksumItem(ProxyRepository proxy,
         throw new LocalStorageException("Null item repository attributes");
       }
 
-      if (Boolean.parseBoolean(attributes.get(noattrname)) && !request.isRequestAsExpired()) {
-        throw new ItemNotFoundException(request);
+      // Check if checksum path is in NFC
+      if (proxy.getNotFoundCache().contains(checksumRequest.getRequestPath()) && !checksumRequest.isRequestAsExpired()) {
```
Comment (cstamas) [inline components/nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/ChecksumContentValidator.java]: Maybe `proxy.maintainNotFoundCache(checksumRequest)` instead?

File: components/nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/ChecksumContentValidator.java [hunk 2]
```diff
@@ -194,37 +184,32 @@ private static RemoteHashResponse doRetrieveChecksumItem(ProxyRepository proxy,
...
-        catch (ItemNotFoundException e) {
-          // fall through
-        }
-        catch (RemoteAccessException e) {
-          // fall through
+          StorageFileItem checksumItem =
+              (StorageFileItem) proxy.getRemoteStorage().retrieveItem(proxy, checksumRequest, proxy.getRemoteUrl());
+
+          // fetched checksum item, extract hash and store it in attributes
+          hash = doStoreChechsumItem(proxy, artifact, attrname, checksumItem);
```
Comment (cstamas) [inline components/nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/ChecksumContentValidator.java]: Probably needs a `proxy.removeFromNotFoundCache(checksumRequest)` to evict it from NFC

**Comment by cstamas** [comment]
+1

**Comment by adreghiciu** [comment]
+1

**Comment by jdillon** [comment]
http://bamboo.s/browse/NX-OSSF628-2 :white_check_mark:

**Comment by mcculls** [comment]
+1 looks good to me

File: components/nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/ChecksumContentValidator.java [hunk 3]
```diff
@@ -235,49 +221,68 @@ private static RemoteHashResponse doRetrieveChecksumItem(ProxyRepository proxy,
...
+  /**
+   * Store checksum information for given artifact in attributes.
+   *
+   * @param proxy         Proxy repository where artifact lives.
+   * @param artifact      Artifact storage-item.
+   * @param attrname      Checksum attribute name.
+   * @param checksumItem  Checksum storage-item.
+   * @return              Checksum hash, read from checksum storage-item.
+   */
+  private static String doStoreChechsumItem(final ProxyRepository proxy,
```
Comment (peterlynch) [inline components/nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/ChecksumContentValidator.java]: minor: doStoreChechsumItem -> doStoreChecksumItem ?

**Comment by rseddon** [comment]
I've been testing this, and what I've found is that it solves the case where the maven-metadata.xml.sha1 file is cached as not existing.  

However, it seems there is still a timing problem here.   Start with the case where both the maven-metadata.xml and maven-metadata.xml.sha1 already exist.  Publish a new maven-metadata.xml file, and retrieve it through a proxy before the new maven-metadata.xml.sha1 file is published.    The previous checksum will be cached, and after this it seems there is no way to get Nexus to check for the updated checksum file.  Expire cache does not work, the old checksum remains cached in the proxy attributes.

**Comment by jdillon** [comment]
This is on-hold for now, given there is still an issue with the proxy-cache, since the checksum "files" aren't really their own storage-items and thus do not have their own last-accessed attributes and can not participate in normal isOld logic that proxy repositories have.

**Comment by jdillon** [comment]
replaced by other changes

**Pull request closed without merging**